src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/HashAlgorithmName.cs

```diff
@@ -104,5 +104,45 @@ public override int GetHashCode()
         {
             return !(left == right);
         }
+
+        public static bool TryFromOid(string oidValue, out HashAlgorithmName value)
+        {
+            switch (oidValue)
+            {
+                case Oids.Md5:
+                    value = MD5;
+                    return true;
+                case Oids.Sha1:
+                    value = SHA1;
+                    return true;
+                case Oids.Sha256:
+                    value = SHA256;
+                    return true;
+                case Oids.Sha384:
+                    value = SHA384;
+                    return true;
+                case Oids.Sha512:
+                    value = SHA512;
+                    return true;
+                default:
+                    value = default;
+                    return false;
+            }
+        }
+
+        public static HashAlgorithmName FromOid(string oidValue)
+        {
+            if (oidValue is null)
+            {
+                throw new ArgumentNullException(nameof(oidValue));
```

Comment:
Why is this in the From and not the Try?

Said another way: The guidance for Try pattern methods is (loosely) "every false case should have the same recovery action".  Do you believe that all of a) null input, b) not a dotted-decimal OID, c) not a known hash algorithm are the same recovery action?  What is it?

(Probably they're all "throw a domain-specific exception", so "yes" is fine; but since null can never work there's an argument that it's a precondition check and should still throw ANE from the Try method)

Reply:
Hm. I had two thoughts.

`TryParse` on int and friends accepts null. While that is apples to oranges a bit, I think some consistency there is useful since those are the most known examples of "try". Generally when folks see "try", that gets simplified to "does not throw".

My use case is simply "the output of another function was invalid as this function's input". For me, `null` and `"scooby doo"` are equally invalid.

That said, I could see an ANE being useful to indicate "maybe you are passing in something uninitialize or want to provide a sensible default (if such a thing exists)"

I don't feel particularly strong about it. I'm happy to throw if that seems more sensible.

Reply:
Let's go ahead and make the Try throw the ANE, too.  When nullability analysis catches up to this DLL it just means we'll say it takes non-null input.

I'd say this is more like Dictionary.TryGetValue (ANE) than int.TryParse (return false).

Reply:
@bartonjs Is it generally expected that every public API validates the parameters on their own, or, if a method is simply delegating to another public API method, can it depend on the delegated method to validate parameters?

Put another way, should both `FromOid` and `TryFromOid` have the `oidValue is null` check?

Reply:
If it's the same parameter name then passing it through like you did is perfectly fine.